src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,432 @@
+     * Delays used to setDelay()
+     */
+    private static final int DELAY_TWENTY = 20;
+    private static final int DELAY_FOURTY = 40;
+    private static final int DELAY_ONE_HUNDRED = 100;
+
+    /**
+     * Number of actions needed to dump inventory.
+     */
+    private static final int ACTIONS_FOR_DUMP = 10;
+
+    /**
+     * Creates the abstract part of the AI.
+     * Always use this constructor!
+     *
+     * @param job the job to fulfill
+     */
+    public AbstractEntityAIHerder(@NotNull final J job, final int maxAnimalsMultiplier)
+    {
+        super(job);
+        this.maxAnimalMultiplier = maxAnimalsMultiplier;
+        super.registerTargets(
+          new AITarget(IDLE, START_WORKING),
+          new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+          new AITarget(PREPARING, this::prepareForHerding),
+          new AITarget(HERDER_DECIDE, this::decideWhatToDo),
+          new AITarget(HERDER_BREED, this::breedAnimals),
+          new AITarget(HERDER_BUTCHER, this::butcherAnimals)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return ACTIONS_FOR_DUMP;
+    }
+
+    /**
+     * Decides what job the herder should switch to, breeding, butchering, or shearing (If Shepherd).
+     *
+     * @return the next AIState the shepherd should switch to, after executing this method.
+     */
+    public AIState decideWhatToDo()
+    {
+        setDelay(DELAY_FOURTY);
+
+        final List<T> animals = new ArrayList<>(getAnimals());
+
+        if (animals.isEmpty())
+        {
+            setDelay(DELAY_ONE_HUNDRED);
+            return HERDER_DECIDE;
+        }
+
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.deciding"));
+
+        final int numOfBreedableAnimals = animals.stream().filter(animal -> animal.getGrowingAge() == 0).toArray().length;
+
+        if (maxAnimals())
+        {
+            return HERDER_BUTCHER;
+        }
+        else if (numOfBreedableAnimals >= NUM_OF_ANIMALS_TO_BREED)
+        {
+            return HERDER_BREED;
+        }
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Redirects the herder to their building.
+     *
+     * @return the next state.
+     */
+    private AIState startWorkingAtOwnBuilding()
+    {
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.goingToHut"));
+        if (walkToBuilding())
+        {
+            return getState();
+        }
+        return PREPARING;
+    }
+
+    /**
+     * Prepares the herder for herding
+     *
+     * @return the next AIState
+     */
+    private AIState prepareForHerding()
+    {
+        if (checkForToolOrWeapon(ToolType.AXE) && checkOrRequestItems(new ItemStack(Items.WHEAT, 2))
+              && (this instanceof EntityAIWorkShepherd && checkForToolOrWeapon(ToolType.SHEARS)))
+        {
+            return getState();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 20 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)